### PR TITLE
dex: configure gRPC and generate the certificates (CA/server/client) during deployment

### PR DESCRIPTION
**Pull request by ccojocar**
In addition, the web cert values were refatored to be consistent with gRPC values.

**What this PR does / why we need it**:

`dex chart`: configure gRPC and generate the certificates (CA/server/client) during deployment

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: stable/dex/templates/job-grpc-certs.yaml [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+{{ $caName := .Values.certs.grpc.secret.caName }}
+{{ $local := dict "i" 0 }}
+apiVersion: batch/v1
+kind: Job
+metadata:
+  annotations:
+    "helm.sh/hook": post-install
+    "helm.sh/hook-weight": "2"
+    "helm.sh/hook-delete-policy": hook-succeeded
+  name: dex-grpc-certs
```
Comment (unguiculus) [inline stable/dex/templates/job-grpc-certs.yaml]: Resource names should be based on the fullname template.

File: stable/dex/templates/job-web-certs.yaml [hunk 2]
```diff
@@ -1,25 +1,26 @@
...
 {{ $local := dict "i" 0 }}
 apiVersion: batch/v1
 kind: Job
 metadata:
   annotations:
     "helm.sh/hook": post-install
+    "helm.sh/hook-weight": "1"
     "helm.sh/hook-delete-policy": hook-succeeded
-  name: {{ template "dex.fullname" . }}
+  name: dex-web-certs
```
Comment (unguiculus) [inline stable/dex/templates/job-web-certs.yaml]: Resource names should be based on the fullname template.

File: stable/dex/values.yaml [hunk 3]
```diff
@@ -13,39 +13,55 @@ nodeSelector: {}
 
 replicas: 1
 
+resources:
```
Comment (unguiculus) [inline stable/dex/values.yaml]: Don't specify defaults for resources. Comment them out.

File: stable/dex/templates/config-openssl.yaml [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{{- if .Values.certs.grpc.create }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  labels:
+    app: {{ template "dex.name" . }}
+    chart: {{ template "dex.chart" . }}
+    heritage: "{{ .Release.Service }}"
+    release: "{{ .Release.Name }}"
+  name: dex-openssl-config
```
Comment (unguiculus) [inline stable/dex/templates/config-openssl.yaml]: Resource names should be based on the fullname template.

**Review by unguiculus — CHANGES_REQUESTED** [review]

**Comment by ccojocar** [comment]
@unguiculus Thanks! I fixed all review comments.

**Comment by ccojocar** [comment]
@unguiculus Pease could you have a look at this PR? Is there anything left to get it merged?

File: stable/dex/values.yaml [hunk 5]
```diff
@@ -61,40 +77,33 @@ serviceAccount:
...
   connectors:
-#  - type: github
-#    id: github
-#    name: GitHub
-#    config:
-#      clientID: xxxxxxxxxxxxxxx
-#      clientSecret: yyyyyyyyyyyyyyyyyyyyy
-#      redirectURI: https://dex.minikube.local:5556/callback
-#      org: kubernetes
+  - type: github
```
Comment (kfox1111) [inline stable/dex/values.yaml]: This actually changes a hint to enabled config. Not sure this is desirable.

File: stable/dex/values.yaml [hunk 6]
```diff
@@ -61,40 +77,33 @@ serviceAccount:
...
+    config:
+      clientID: clientID
+      clientSecret: clientSecret
+      redirectURI: https://dex.io/callback
+      orgs:
+        - name: jenkins-x
   oauth2:
     skipApprovalScreen: true
 
-#  staticClients:
```
Comment (kfox1111) [inline stable/dex/values.yaml]: This was a useful example of how to use this. Please put it back.

**Review by kfox1111 — CHANGES_REQUESTED** [review]

**Comment by ccojocar** [comment]
@unguiculus @kfox1111 I pushed the changes for review.

**Comment by kfox1111** [comment]
lgtm

**Review by unguiculus — APPROVED** [review]

**Comment by andrewrynhard** [comment]
@ccojocar What is the recommended way to use the cert generation feature? There is sort of a chicken and egg problem here is there not? You need to have the `--oidc-ca-file` to the API server set to use the web CA secret generated, but you don't have that until you deploy this chart.

**Comment by ccojocar** [comment]
@andrewrynhard Are you talking about web endpoint CA or gRPC CA? Where do you need to set the `--oidc-ca-file` flag?

The job which generates the self-signed certificate for the web endpoint was there before my changes. It basically creates a self-signed certificate which is stored in a secret along with the CA. The public certificate of the CA is in addition stored in a configmap, which you can mount in a POD.

**Pull request merged**